ably/realtime_channel.go

```diff
@@ -411,6 +452,45 @@ func (c *RealtimeChannel) notify(msg *proto.ProtocolMessage) {
 	}
 }
 
+func (c *RealtimeChannel) lockStartRetryAttachLoop(err error) {
+	// TODO: Move to SUSPENDED; move it to DETACHED for now.
+	c.state.set(StateChanDetached, err)
+	c.state.Unlock()
+
+	go func() {
+		// TODO: The listener should be set before unlocking the state.
+		// Will do once we have an EventEmitter.
+		stateChange := make(chan State, 10)
+		c.state.on(stateChange)
+		defer c.state.off(stateChange)
+
+		for !c.retryAttach(stateChange) {
+		}
+	}()
+}
+
+func (c *RealtimeChannel) retryAttach(stateChange chan State) (done bool) {
+	select {
+	case <-c.opts().After(c.opts().ChannelRetryTimeout):
```

Comment:
`c.opts().After` is a callback. There is no default value for this set. It is passing in your test because you are explicitly setting this , this part will panic at runtime.

We should also avoid `time.After` according to docs https://pkg.go.dev/time?tab=doc#After  it might result into leaks in the case we receive  any concurrent changes that cancels retry.